core/src/transport/mod.rs

```diff
@@ -54,6 +54,24 @@ pub use self::memory::connector;
 pub use self::muxed::MuxedTransport;
 pub use self::upgrade::UpgradedNode;
 
+
+#[derive(Debug, Fail)]
+pub enum TransportError {
+    #[fail(display = "Listening is not support for Multiaddress {}", _0)]
+    ListenNotSupported(Multiaddr),
+
+    #[fail(display = "Listening on Multiaddress {} failed: {}", _0, _1)]
+    ListenFailed(Multiaddr, #[cause] IoError),
+
+    #[fail(display = "Dialing is not support for Multiaddress {}", _0)]
+    DialNotSupported(Multiaddr),
+
+    #[fail(display = "Dialing for Multiaddress {} failed: {}", _0, _1)]
+    DialingFailed(Multiaddr, #[cause] IoError),
+}
```

Comment:
I am not set on using `failure`, I was just doing it in the spirit of building for the future ;) . Though, I am not too afraid that `failure` will change significantly, we also don't use any of its particular features at the moment. However, we might want to start thinking about a good `Error`-story in libp2p, as I feel, at the moment, often errors are just logged or passed on in a very generic way, rather than something explicit that a higher layer could catch and handle...

Reply:
Errors should definitely be improved, but in my opinion we should just stick to manually-defined errors or IoError for the sake of consistency. Having one single crate in the repo use `failure` is not great.

Reply:
replaced with a custom error type, dropped `failure`-crate.